src/core/net/ip6.cpp

```diff
@@ -603,11 +614,303 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+    }
+
+    // copy the fragment payload into the message buffer
+    assertValue = aMessage.CopyTo(aMessage.GetOffset() + sizeof(fragmentHeader), aMessage.GetOffset() + offset,
+                                  payloadFragment, *message);
+    assert(assertValue == static_cast<int>(payloadFragment));
+
+    // check if it is the last frame
+    if (fragmentHeader.IsMoreFlagSet() == false)
+    {
+        // use the offset value for the whole ip message length
+        SuccessOrExit(error = message->SetOffset(offset + payloadFragment + aMessage.GetOffset()));
+    }
+
+    if (message->GetOffset() >= message->GetLength())
+    {
+        offset = 0;
+
+        // creates the header for the reassembled ipv6 package
+        VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+        header.SetPayloadLength(message->GetLength() - sizeof(header));
+        header.SetNextHeader(static_cast<IpProto>(fragmentHeader.GetNextHeader()));
+        assertValue = message->Write(0, sizeof(header), &header);
+        assert(assertValue == sizeof(header));
+
+        otLogDebgIp6("Reassembly complete.");
+
+        mReassemblyList.Dequeue(*message);
+
+        error = HandleDatagram(*message, aNetif, aMessageInfo.mLinkInfo, aFromNcpHost);
+    }
+
+exit:
+    if (error != OT_ERROR_DROP && error != OT_ERROR_NONE && isFragmented)
+    {
+        if (message != NULL)
+        {
+            mReassemblyList.Dequeue(*message);
+            message->Free();
+        }
+        otLogWarnIp6("Reassembly failed: %s", otThreadErrorToString(error));
+    }
+    if (isFragmented)
+    {
+        // drop all fragments, the payload is stored in the fragment buffer
+        error = OT_ERROR_DROP;
+    }
+
+    return error;
+}
+
+void Ip6::CleanupFragmentationBuffer(void)
+{
+    Message *message;
+    Message *next;
+
+    for (message = mReassemblyList.GetHead(); message; message = next)
+    {
+        next = message->GetNext();
+        mReassemblyList.Dequeue(*message);
+
+        message->Free();
+    }
+}
+
+void Ip6::HandleTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<Ip6>().HandleUpdateTimer();
+}
+
+void Ip6::HandleUpdateTimer(void)
+{
+    if (UpdateReassemblyList())
+    {
+        mTimer.Start(kStateUpdatePeriod);
+    }
+}
+
+bool Ip6::UpdateReassemblyList(void)
+{
+    Message *next;
+
+    for (Message *message = mReassemblyList.GetHead(); message; message = next)
+    {
+        next = message->GetNext();
+
+        if (message->GetTimeout() > 0)
+        {
+            message->DecrementTimeout();
+        }
+        else
+        {
+            otLogNoteIp6("Reassembly timeout.");
+            SendIcmpError(*message, IcmpHeader::kTypeTimeExceeded, IcmpHeader::kCodeFragmReasTimeEx);
+
+            mReassemblyList.Dequeue(*message);
+            message->Free();
+        }
+    }
+    return mReassemblyList.GetHead() != NULL;
```

Comment:
nit: given the method name, prefer that this method return `void` and move this check into `Ip6::HandleUpdateTimer()`.

Reply:
changed